Add generalized longitudinal job

fbertsch opened the pull request:
This will allow us to make Longitudinal-type datasets out of any parquet datasets. I'll be using this for the mobile ones to begin with. With Presto now supporting lambda-expressions, these datasets will be easy-to-use in STMO.

The basic idea here is to give it a dataset name, a grouping column, and any number of ordering columns. Most datasets will be grouped by client_id, and ordered by submission_date or activity_date.

Ideas for more test cases are welcome!

fbertsch commented:
@harterrt I figured out a way to make it *wayyy* simpler :)

harterrt commented:
Cool, should I hold off on reviewing for a while then?

fbertsch commented:
@harterrt nope, the simpler version is the current version. Carry on!

harterrt commented on src/main/scala/com/mozilla/telemetry/utils/udfs.scala:
```diff
@@ -0,0 +1,150 @@
+
+  private def sort(order: Seq[Int], orderType: DataType, seq: Seq[Any]): Seq[Int] = {
+    orderType match {
+      case _: IntegerType => order.sortWith(sortInt(seq.asInstanceOf[Seq[Integer]]))
```
Do we need an Option here?

fbertsch replied:
Nope, Integer allows for null values.

harterrt commented on src/test/scala/com/mozilla/telemetry/views/GeneralizedLongitudinalTest.scala:
```diff
@@ -0,0 +1,144 @@
+
+  "longitudinal" must "correctly aggregate strings" in {
+    val oss = fixture.longitudinal.map(_.getSeq(fixture.fieldIndex("os")).toList).toList
+    assert(oss == List(List("nothing", "Android", "iOS"), List("Android", null)))
```
Shouldn't the OS list for client 2 be : `List(null, "Android")`?

fbertsch replied:
It's sorting by `submission_date_s3`, which has `"20160101"` for `"Android"`, and `null` for `null`. The `null` value is sorted last, so the OS list is `List("Android", null)`.

harterrt replied:
As discussed offline, I (incorrectly) thought this was ordered by `order`. No change needed.

harterrt commented on src/test/scala/com/mozilla/telemetry/views/GeneralizedLongitudinalTest.scala:
```diff
@@ -0,0 +1,144 @@
+    assert(numNulls == 2)
+  }
+
+  "longitudinal" must "sort null last" in {
```
This seems wrong to me

fbertsch replied:
So "Sort null last" means that, for a give sorting column[s], null values will always be last. So in your example, the results would be as follows:

## Sorting by data_a
grouping | data_a | data_b
--------- | ------- | ------
a | [6, null] | [null, 9]

## Sorting by data_b
grouping | data_a | data_b
--------- | ------- | ------
a | [null, 6] | [9, null]

fbertsch replied:
For reference, see the big "SORTING" block comment in the CollectList class.

harterrt replied:
As with above, you're right. I didn't realize this was ordered by `submission_date` and not by `order`.

harterrt commented on src/main/scala/com/mozilla/telemetry/utils/udfs.scala:
```diff
@@ -0,0 +1,150 @@
+   * case e1 < e2 => e2 before e1
+   **/
+
+  private def sortString(arr: Seq[String])(e0: Int, e1: Int): Boolean = {
```
Do we have any datasets where we would want to sort by an `Option[Int]`?

fbertsch replied:
Right now both Integer and String allow null values. Some others - such as Long - don't, so we're going to have to deal with an Option[Long] for that sorting case.

fbertsch commented on src/main/scala/com/mozilla/telemetry/utils/udfs.scala:
```diff
@@ -0,0 +1,150 @@
+      case _: BooleanType   => sortTrimTypedArray[Boolean](buffer.getSeq(arrayIndex), sortOrder)
+      case _: TimestampType => sortTrimTypedArray[Timestamp](buffer.getSeq(arrayIndex), sortOrder)
+      case _: DateType      => sortTrimTypedArray[Date](buffer.getSeq(arrayIndex), sortOrder)
+      case _: ArrayType     => sortTrimTypedArray[Seq[Any]](buffer.getSeq(arrayIndex), sortOrder)
```
Maps seem to work, based on the test I have. I can add another if you think the specific check is not covered.

harterrt replied:
Missed that test. This is good to go.

fbertsch commented:
@harterrt answered most queries above, made some minor changes. RFAL.

fbertsch commented:
@harterrt made a few changes:
1. Added sorting for Long type
2. Made sorting of arrays O(n) rather than O(nlog(n))
3. Fixed secondary ordering test
4. Sorting functions return false if the values are equal - per sorting spec

sunahsuh commented:
Commenting to register my interest in using this for experiments. For my needs, I'll need multiple grouping columns (`client_id`, `experiment_name`, `experiment_branch`) but I'm happy to make the change to allow that after this is merged.
